### PR TITLE
1050: PCIe Topology Cable Links (#338)

### PR DESCRIPTION
**Pull request by baemyung**
This commit adds a downstream/upstream port, downstream/upstream resource, and downstream/upstream chassis link into the cable JSON response, which represents an association between resources associated via cables for williwakas and cable cards.

Upstream: https://gerrit.openbmc.org/c/openbmc/bmcweb/+/61122

(More discussion is in https://github.com/ibm-openbmc/dev/issues/3533)

Tested using a rainier system and run Redfish-Service-Validator/ both pass

```
$ curl -k -H "X-Auth-Token: $token" -X GET https://${bmc}/redfish/v1/Cables/dp0_cable0
{
  "@odata.id": "/redfish/v1/Cables/dp0_cable0",
  "@odata.type": "#Cable.v1_2_0.Cable",
  "CableStatus": "Normal",
  "CableType": "",
  "Id": "dp0_cable0",
  "Links": {
    "DownstreamPorts": [
      {
        "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector1"
      }
    ],
    "DownstreamResources": [
      {
        "@odata.id": "/redfish/v1/Chassis/chassis/Assembly#/Assemblies/24"
      }
    ],
    "UpstreamPorts": [
      {
        "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/pcie_card10/Ports/c10_connector0"
      }
    ],
    "UpstreamResources": [
      {
        "@odata.id": "/redfish/v1/Systems/system/PCIeDevices/pcie_card8"
      },
      {
        "@odata.id": "/redfish/v1/Systems/system/PCIeDevices/pcie_card10"
      }
    ]
  },
  "Name": "Cable",
  ...
}
```

### CODE REVIEW

File: redfish-core/lib/cable.hpp [hunk 1]
```diff
@@ -3,11 +3,64 @@
...
+        if ((*value).empty())
+        {
+            BMCWEB_LOG_DEBUG << "No association found for cable";
+            messages::internalError(asyncResp->res);
+            return;
+        }
+        callback(*value);
+        },
+        "xyz.openbmc_project.ObjectMapper", cableObjectPath,
+        "org.freedesktop.DBus.Properties", "Get",
```
Comment (gtmills) [inline redfish-core/lib/cable.hpp:60]: Upstream we are using 
        sdbusplus::asio::getProperty
There is also a utility function you could pull in at https://gerrit.openbmc.org/c/openbmc/bmcweb/+/60872 
Up to you if you want to move

File: redfish-core/lib/cable.hpp [hunk 2]
```diff
@@ -3,11 +3,64 @@
...
+ * @param[in]       callback        callback method
+ */
+template <typename Callback>
+inline void
+    linkAssociatedCable(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                        const std::string& cableObjectPath, Callback&& callback)
+{
+    crow::connections::systemBus->async_method_call(
+        [asyncResp, callback{std::forward<Callback>(callback)}](
+            const boost::system::error_code ec,
```
Comment (gtmills) [inline redfish-core/lib/cable.hpp:26]: include what you use? #include <boost/system/error_code.hpp>

File: redfish-core/lib/cable.hpp [hunk 3]
```diff
@@ -3,11 +3,64 @@
...
+
+/**
+ * @brief Api to get Cable Association.
+ * @param[in,out]   asyncResp       Async HTTP response.
+ * @param[in]       cableObjectPath Object path of the Cable with association.
+ * @param[in]       callback        callback method
+ */
+template <typename Callback>
+inline void
+    linkAssociatedCable(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
```
Comment (gtmills) [inline redfish-core/lib/cable.hpp:21]: #include <memory> 
#include <string>

Maybe run the https://include-what-you-use.org/ ?

File: redfish-core/lib/cable.hpp [hunk 4]
```diff
@@ -3,11 +3,64 @@
...
+                // This cable have no association.
+                BMCWEB_LOG_DEBUG << "No association found";
+                return;
+            }
+            BMCWEB_LOG_ERROR << "DBUS response error" << ec.message();
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        const std::vector<std::string>* value =
```
Comment (gtmills) [inline redfish-core/lib/cable.hpp:41]: #include <vector>

File: redfish-core/lib/cable.hpp [hunk 5]
```diff
@@ -3,11 +3,64 @@
...
+            std::get_if<std::vector<std::string>>(&(endpoints));
```
Comment (gtmills) [inline redfish-core/lib/cable.hpp:42]: #include <variant>

File: redfish-core/lib/cable.hpp (hunk 2 above)
Comment (baemyung) [reply redfish-core/lib/cable.hpp:26]: Thanks.
It looks like I need to compile.  I'll do it later.
BTW, is this perhaps better to have in pool machine - e.g. gfwa826?

File: redfish-core/lib/cable.hpp (hunk 1 above)
Comment (baemyung) [reply redfish-core/lib/cable.hpp:60]: I've changed to use dbus::utility::getAssociationEndPoints().

**Comment by baemyung** [comment]
@gtmills 
Thanks for review. I've changed the code and verified it too.

**Review by gtmills — APPROVED** [review]

**Pull request merged**